otelcol/internal/stefexporter/internal/connmanager.go

```diff
@@ -0,0 +1,420 @@
+// ConnManager needs.
+type ManagedConn struct {
+	Conn       Conn
+	startTime  time.Time
+	lastFlush  time.Time
+	needsFlush bool
+	isAcquired bool
+}
+
+// ConnCreator allow creating connections.
+type ConnCreator interface {
+	// Create a new connection. May be called concurrently.
+	// The attempt to create the connection should be cancelled if ctx is done.
+	Create(ctx context.Context) (Conn, error)
+}
+
+// Conn represents a connection that can be closed or flushed.
+type Conn interface {
+	// Close the connection. The connection will be discarded
+	// after this call returns.
+	Close(ctx context.Context) error
+
+	// Flush the connection. This is typically to send any buffered data.
+	// Will be called periodically (see ConnManager flushPeriod) and
+	// before ConnManager.Stop returns.
+	Flush() error
+}
+
+func NewConnManager(
+	logger *zap.Logger,
+	creator ConnCreator,
+	targetConnCount uint,
+	flushPeriod time.Duration,
+	reconnectPeriod time.Duration,
+) *ConnManager {
+	return &ConnManager{
+		logger:          logger,
+		clock:           clockwork.NewRealClock(),
+		connCreator:     creator,
+		targetConnCount: targetConnCount,
+		idleConns:       make(chan *ManagedConn, targetConnCount),
+		recreateConns:   make(chan *ManagedConn, targetConnCount),
+		flushPeriod:     flushPeriod,
+		reconnectPeriod: reconnectPeriod,
+		stoppedCond:     NewCancellableCond(),
+		stopSignal:      make(chan struct{}),
+	}
+}
+
+// Start starts the connection manager. It will immediately start
+// creating targetConnCount new connections. All successfully
+// created connections will become available for acquisition.
+func (c *ConnManager) Start() {
+	// Create connections in recreateConns pool. They will be created
+	// by recreator.
+	for i := uint(0); i < c.targetConnCount; i++ {
+		c.recreateConns <- &ManagedConn{}
+		c.curConnCount.Add(1)
+	}
+
+	go c.flusher()
+	go c.durationLimiter()
+	go c.recreator()
+}
+
+// Stop stops the connection manager. It will wait until all acquired
+// connections are returned. Then it will flush connections that
+// are marked as needing to flush, and then will close all connections.
+func (c *ConnManager) Stop(ctx context.Context) error {
+	// Signal goroutines to stop
+	close(c.stopSignal)
+
+	// Wait until all goroutines stop
+	err := c.stoppedCond.Wait(
+		ctx, func() bool {
+			return c.flusherStopped && c.durationLimiterStopped && c.recreatorStopped
+		},
+	)
+	if err != nil {
+		c.logger.Error("Failed to stop connection manager", zap.Error(err))
+		return err
+	}
+
+	return c.closeAll(ctx)
+}
+
+// Close all connections.
+func (c *ConnManager) closeAll(ctx context.Context) error {
+	// We must close exactly targetConnCount connections in total.
+	// All goroutines are stopped at this point, so they won't interfere.
+	// All connections are either in the between idleConns, flushConns and recreateConns
+	// pools or are acquired and will be returned to one of the pools soon.
+
+	cnt := c.curConnCount.Load()
```

Comment:
Why do we have to track `curConnCount`? Can't we use `targetConnCount` instead here?

Reply:
We need to close as many connections as we created minus what is already closed, which is tracked in `curConnCount`). `curConnCount` may be lower than `targetConnCount` if we didn't manage yet to create all the necessary connections.